File: examples/ts-guess/ts/logic.ts
export class GuessingGame {
    lower: number;
    upper: number;
    target: number | null;

    constructor(lower = 1, upper = 100) {
        this.lower = lower;
        this.upper = upper;
        this.target = null;
    }

    start(who, lower?: number, upper?: number) {
        this.lower = lower || this.lower;
        this.upper = upper || this.upper;
        this.target = math.random(this.lower, this.upper);
        print(`Guess a number between ${this.lower} and ${this.upper} (inclusive!)`)
    }

    finish(who) {
        print(`${who.steam_name} won the game!`);
        this.target = null;
    }

    terminate(who) {
        print(`${who.steam_name} terminated the game :(`);
        this.target = null;
    }

    guess(message: string, who) {
        if (this.target == null) {
            return;
        }

        const guess = tonumber(message)
        if (guess == null) {
            return;
        }

        if (guess == this.target) {
            this.finish(who);
        } else if (guess < this.target) {
            print("Higher!");
        } else {
            print("Lower!");
        }
    }
}

File: examples/lua-guess/lua/global/logic.lua
local logic = {}

-- functions for game logic
function logic.start(who, new_lower, new_upper)
    upper = new_upper or 100
    lower = new_lower or 1
    target = math.random(lower, upper)
    print("Guess a number between " .. lower .. " and " .. upper .. " (inclusive!)")
end

function logic.finish(who)
    print(who.steam_name .. " won the game!")
    target = nil
end

function logic.terminate(who)
    print(who.steam_name .. " terminated the game :(")
    target = nil
end

function logic.handleGuess(message, sender)
    -- ignore messages when there's no active game
    if target == nil then
        return
    end

    -- check `message` as `guess` against `target`
    local guess = tonumber(message)

    -- ignore non-numeric messages
    if guess == nil then
        return
    end

    -- respond to the player
    if guess == target then
        logic.finish(sender)
    elseif guess < target then
        print("Higher!")
    else
        print("Lower!")
    end
end

return logic